scripts/usb_scan.py

```diff
@@ -0,0 +1,618 @@
+    _PATH = "/etc/xensource/usb-policy.conf"
+
+    _CLASS = "class"
+    _SUBCLASS = "subclass"
+    _PROTOCOL = "prot"
+    _ID_VENDOR = "vid"
+    _ID_PRODUCT = "pid"
+    _BCD_DEVICE = "rel"
+
+    # key in policy <--> key in usb device
+    _KEY_MAP_DEVICE = {_ID_VENDOR: UsbDevice._ID_VENDOR,
+                       _ID_PRODUCT: UsbDevice._ID_PRODUCT,
+                       _BCD_DEVICE: UsbDevice._BCD_DEVICE}
+
+    # key in policy <--> key in usb interface
+    _KEY_MAP_INTERFACE = {_CLASS: UsbInterface._CLASS,
+                          _SUBCLASS: UsbInterface._SUB_CLASS,
+                          _PROTOCOL: UsbInterface._PROTOCOL}
+
+    _PAT_KEY = r"\s*({}|{}|{}|{}|{}|{})\s*".format(_CLASS, _SUBCLASS,
+                                                   _PROTOCOL, _ID_VENDOR,
+                                                   _ID_PRODUCT, _BCD_DEVICE)
+    _PATTERN = r"{}=\s*([0-9a-f]+)".format(_PAT_KEY)
+
+    _ALLOW = "allow"
+
+    def __init__(self):
+        """ parse policy file, generate rule list
+
+        Note: hubs are never allowed to pass through
+        """
+        self.rule_list = []
+        try:
+            with open(self._PATH, "r") as f:
+                log_wrapper("=== policy file begin")
+                for line in f:
+                    log_wrapper(line[0:-1])
+                    self.parse_line(line)
+                log_wrapper("=== policy file end")
+        except IOError:
+            # without policy file, no device will be allowed to passed through
+            log_wrapper(traceback.format_exc())
+
+        log_wrapper("=== rule list begin")
+        log_list_wrapper(self.rule_list)
+        log_wrapper("=== rule list end")
+
+    def parse_line(self, line):
+        """ parse one line of policy file, generate rule, and append it to
+        self.rule_list
+
+        Example:
+        ALLOW:vid=056a pid=0315 class=03 # Wacom Intuos tablet
+            => {'pid': '0315', 'vid': '056a', 'allow': True, 'class': '03'}
+        DENY: class=03 subclass=01 prot=01 # HID Boot keyboards
+            => {'prot': '01', 'class': '03', 'allow': False, 'subclass': '01'}
+        ALLOW: # Otherwise allow everything else
+            => {'allow': True}
+
+        :param line: (str) single line of policy file
+        :return: None
+        """
+        # 1. remove comments
+        # ^([^#]*)(#.*)?$
+        i = line.find("#")
+        if i != -1:
+            line = line[0:i]
+
+        # 2. split action and match field
+        # ^\s*(ALLOW|DENY)\s*:\s*([^:]*)$
+        if ":" not in line:
+            return
+        parts = line.split(":")
+        n = len(parts)
+        if n == 0 or n > 2:
+            return
+
+        # 3. parse action
+        # \s*(ALLOW|DENY)\s*
+        rule = {}
+        action = parts[0].strip()
+        if action.lower() == "allow":
+            rule[self._ALLOW] = True
+        elif action.lower() == "deny":
+            rule[self._ALLOW] = False
+        else:
+            return
```

Comment:
Log this case with an explicative message and at error level. At this stage we are expecting the rules to be well formed.

Reply:
Done.